Mobile/cellular broadband support

davidmhewitt opened the pull request:
Fixes #2 (but doesn't close the bounty yet, as that refers to the switchboard plug too, which I'll do next!)

I don't know if anyone has the hardware to test this, but it functions as you'd expect with any cellular modem supported by NetworkManger and/or ModemManager. You can connect/disconnect using the toggle switch and it shows the signal strength in the icon. Can provide screenshots/screencasts if you don't believe me ;)

I've written the code in such a way that WiFi networks and wired networks take priority over the cellular network, so if you have either of those connected, they show their status icon in the panel rather than the cellular modem.

matzipan commented:
I have some hardware to test this. Will do.

davidmhewitt commented:
Just keep in mind that you need to set up the connection with the network manager GUI via the "edit connections" button in the network plug since I havent done the switchboard part of this yet.

matzipan commented:
Ok, so far I can see that there's an entry in the popover for mobile and if I disable wireless, the icon changes to a disabled mobile network. Tomorrow I will check with a sim card and configuration.

donadigo commented on src/Widgets/DisplayWidget.vala:
```diff
@@ -87,6 +87,43 @@ public class Network.Widgets.DisplayWidget : Gtk.Box {
             });
             break;
+        case Network.State.CONNECTED_MOBILE_WEAK:
+            image.icon_name = "network-cellular-signal-weak-%ssymbolic".printf (secure? "secure-" : "");
```
Space between `secure` and `?`.

donadigo commented on src/Widgets/DisplayWidget.vala:
```diff
@@ -87,6 +87,43 @@ public class Network.Widgets.DisplayWidget : Gtk.Box {
+            image.icon_name = "network-cellular-signal-weak-%ssymbolic".printf (secure? "secure-" : "");
+            break;
+        case Network.State.CONNECTED_MOBILE_OK:
+            image.icon_name = "network-cellular-signal-ok-%ssymbolic".printf (secure? "secure-" : "");
```
Same.

donadigo commented on src/Widgets/DisplayWidget.vala:
```diff
@@ -87,6 +87,43 @@ public class Network.Widgets.DisplayWidget : Gtk.Box {
+            image.icon_name = "network-cellular-signal-ok-%ssymbolic".printf (secure? "secure-" : "");
+            break;
+        case Network.State.CONNECTED_MOBILE_GOOD:
+            image.icon_name = "network-cellular-signal-good-%ssymbolic".printf (secure? "secure-" : "");
```
Same.

donadigo commented on src/Widgets/DisplayWidget.vala:
```diff
@@ -87,6 +87,43 @@ public class Network.Widgets.DisplayWidget : Gtk.Box {
+            image.icon_name = "network-cellular-signal-good-%ssymbolic".printf (secure? "secure-" : "");
+            break;
+        case Network.State.CONNECTED_MOBILE_EXCELLENT:
+            image.icon_name = "network-cellular-signal-excellent-%ssymbolic".printf (secure? "secure-" : "");
```
Same.

donadigo commented on src/Widgets/DisplayWidget.vala:
```diff
@@ -87,6 +87,43 @@ public class Network.Widgets.DisplayWidget : Gtk.Box {
+                    strength = "excellent";
+                    break;
+                }
+                image.icon_name = "network-cellular-signal-" + strength + (secure? "secure-" : "")  + "-symbolic";
```
Same.

donadigo commented on src/Widgets/DisplayWidget.vala:
```diff
@@ -87,6 +87,43 @@ public class Network.Widgets.DisplayWidget : Gtk.Box {
+                case 3:
+                    strength = "excellent";
+                    break;
+                }
```
Newline here.

donadigo commented on src/Widgets/DisplayWidget.vala:
```diff
@@ -87,6 +87,43 @@ public class Network.Widgets.DisplayWidget : Gtk.Box {
+            break;
+        case Network.State.CONNECTING_MOBILE:
+            animation_timeout = Timeout.add (300, () => {
+                animation_state = (animation_state + 1) % 4;
```
Can we use declare a different `uint` for this `animation_state` and also for the `animation_timeout`. This could very easily be broke because the same fields are used for WiFi image animation.

donadigo commented on src/Widgets/DisplayWidget.vala:
```diff
@@ -87,6 +87,43 @@ public class Network.Widgets.DisplayWidget : Gtk.Box {
+                    break;
+                }
+                image.icon_name = "network-cellular-signal-" + strength + (secure? "secure-" : "")  + "-symbolic";
+                return true;
```
Explicitly return `true` here makes the timeout run always, even if the state is not `CONNECTING_MOBILE`. I'm not gonna force you to fix this though as this is also the case in master with WiFi.

davidmhewitt commented:
Thanks for the review @donadigo . This is the problem with blindly copying code to copy the style! :stuck_out_tongue: 

Should have addressed all of your issues.

danrabbit commented:
From the screencast this seems to be working as expected. I wonder if we'd want to show both the cellular network signal and the wifi signal at the same time, but this will probably need more user testing to know what the expectation is. So LGTM. Nice work :+1:  :)
